opentracing-impl/src/test/java/io/opentracing/AbstractTracerTest.java

```diff
@@ -89,6 +89,23 @@ public void testNonEmptyExtract() {
         assertEquals("Should find the marker", "whatever", ((TestSpanBuilder)result).operationName);
     }
 
+    @Test
+    public void testExtractAsParent() throws Exception {
+        Map<String,String> map = Collections.singletonMap("test-marker", "whatever");
+        TextMapExtractAdapter adapter = new TextMapExtractAdapter(map);
+        AbstractTracer tracer = new TestTracerImpl();
+        SpanContext parent = tracer.extract(Format.Builtin.TEXT_MAP, adapter);
+        assert AbstractNoopSpan.INSTANCE != tracer.buildSpan("child").asChildOf(parent).start();
+    }
+
+    @Test
+    public void testExtractOfNoParent() throws Exception {
+        AbstractTracer tracer = new TestTracerImpl();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf((Span)AbstractNoopSpan.INSTANCE).start();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf((SpanContext)AbstractNoopSpan.INSTANCE).start();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf(AbstractNoopSpanBuilder.INSTANCE).start();
```

Comment:
>  What I'm saying is this node should always start a trace in this case.

This is what i've referred to as "forced sampling". Where you look for an existing trace ID, and continue sampling if a trace id is found, or force sampling when one doesn't. The exact terminology used here isn't that important.

The default should be not to force sampling. The reasoning about what provides the minimal, most elegant, code.

A) Taking this default the situation of only tracing where a trace exists looks like 

    public Response serverProcessingRequest(Request request) {
        TextMapExtractAdapter adapter = new TextMapExtractAdapter(request.getHeaders());
        SpanContext parent = tracer.extract(Format.Builtin.TEXT_MAP, adapter);
        try (Span span = tracer.buildSpan("some-name").asChildOf(parent).start()) {
            // server processes request
            return response;
        }
    }

B) The tracing from the top level application, where a existing trace if won't ever exist, looks like

    public Response serverProcessingRequest(Request request) {
        try (Span span = tracer.buildSpan("some-name").asChildOf(parent).start()) {
            // server processes request
            return response;
        }
    }

C) And tracing from a service that must trace regardless if a trace exists or not looks like

    public Response serverProcessingRequest(Requestr request) {
        TextMapExtractAdapter adapter = new TextMapExtractAdapter(request.getHeaders());
        SpanContext parent = tracer.extract(Format.Builtin.TEXT_MAP, adapter);
        SpanBuilder builder = tracer.buildSpan("some-name");
        if (!parent instanceof NoopSpanContext) {
            builder = builder.asChildOf(parent);
        }
        try (Span span = builder.start()) {
            // server processes request
            return response;
        }
    }

To begin with (C) is clumsy. I expect a better approach, and I've referred to this as the forced sampling mechanism before. With the idea of tackling that after this PR let's accept it's clumsiness.

The reason this works can be shown when you look at a large enterprise platform. For example if we take 400 services in an enterprise microservices platform. One of those services is a web application: we presume no external traces exist and (B) is applied. Another of those services is a public API that may be used by the web application might also by other external agents: external traces may exist, but tracing is always initialised, so (C) is applied. The remaining 398 services only trace when a trace already exist, so (A) is applied. On top of this there are 50 libraries re-used in various degrees in the services: they have (A) applied as well.

This is just an example but here you see that even if we can reduce the "forced sampling" mechanism down to two lines of code we have still saved ~1K lines of code.

Another way to look at this is that we, doing it this way, make the code we put into libraries and frameworks (that know nothing about the tracing context and implementation, we have absolutely no relationship with where they are deployed) as non-intrusive as possible.